scheds/rust/scx_spark/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1961 @@
+		return true;
+
+	/*
+	 * If the current task is a per-CPU kthread running on the wakee's
+	 * previous CPU, treat it as a synchronous wakeup.
+	 *
+	 * The assumption is that the wakee had queued work for the per-CPU
+	 * kthread, which has now finished, making the wakeup effectively
+	 * synchronous. An example of this behavior is seen in IO
+	 * completions.
+	 */
+	if (is_kthread(current) && (current->nr_cpus_allowed == 1) &&
+	    (prev_cpu == this_cpu))
+		return true;
+
+	return false;
+}
+
+/*
+ * Return the target CPU for @p in case of a sync wakeup.
+ *
+ * During a sync wakeup, the waker commits to releasing the CPU immediately
+ * after the wakeup event, so we should consider a sync wakeup almost like
+ * a direct function call between a waker and a wakee.
+ */
+static s32 try_sync_wakeup(const struct task_struct *p, s32 prev_cpu, s32 this_cpu)
+{
+	/*
+	 * If @prev_cpu is idle, keep using it, since there is no guarantee
+	 * that the cache hot data from the waker's CPU is more important
+	 * than cache hot data in the wakee's CPU.
+	 */
+	if ((this_cpu != prev_cpu) && scx_bpf_test_and_clear_cpu_idle(prev_cpu))
+		return prev_cpu;
+
+	/*
+	 * If waker and wakee are on the same CPU and no other tasks are
+	 * queued, consider the waker's CPU as idle.
+	 */
+	if (!scx_bpf_dsq_nr_queued(SCX_DSQ_LOCAL_ON | this_cpu))
+		return this_cpu;
+
+	return -EBUSY;
+}
+
+/*
+ * Helper function to find idle CPU with specific constraints.
+ */
+static s32 find_idle_cpu_in_mask(const struct cpumask *mask, u64 flags)
+{
+	if (!mask)
+		return -1;
+	return scx_bpf_pick_idle_cpu(mask, flags);
+}
+
+static s32 pick_idle_cpu_gpu(struct task_struct *p, s32 prev_cpu, u64 wake_flags, bool *is_idle)
+{
+	const struct cpumask *big_mask;
+	const struct cpumask *turbo_mask;
+	s32 cpu = -1; //If no idle CPU is found, go down the regular path for finding an idle CPU
+
+	big_mask = cast_mask(big_cpumask);
+	turbo_mask = cast_mask(turbo_cpumask);
+	if (big_mask && bpf_cpumask_empty(big_mask)){
+		big_mask = NULL;
+		turbo_mask = NULL;
+		return cpu;
+	}
+
+	if(turbo_mask && bpf_cpumask_empty(turbo_mask)){
+		turbo_mask = NULL;
+	}
+
+
+/*Try looking for idle turbo CPUs, even if it's not a cache sibling. NOTE: This will probably happen dynamically based on cache sensitivity when workload detection improves.*/
+if(turbo_mask){
+    //Check if the previous CPU is idle and in the turbo mask
+	if (bpf_cpumask_test_cpu(prev_cpu, turbo_mask) && scx_bpf_test_and_clear_cpu_idle(prev_cpu)) {
+		*is_idle = true;
+		return prev_cpu;
+	}
+//Select any turbo CPU :
+	cpu = find_idle_cpu_in_mask(turbo_mask, 0);
+	if(cpu >= 0){
+		*is_idle = true;
+		return cpu;
+	}
+}
+	/*
+	 * Try to re-use the same CPU if it's a big CPU.
+	 */
+	if (big_mask && bpf_cpumask_test_cpu(prev_cpu, big_mask) && scx_bpf_test_and_clear_cpu_idle(prev_cpu)) {
+		*is_idle = true;
+		return prev_cpu;
+	}
+
+	//All big CPUs are L3 cache siblings
+	if (big_mask) {
+		cpu = find_idle_cpu_in_mask(big_mask, 0);
```

Comment:
Ditto for the big mask, it'd be nice (somehow) to prevent cross-LLC migrations (if they can happen).